### PR TITLE
vmspawn: generate predicatable TAP device names and MAC addresses

### PR DESCRIPTION
**Pull request by sam-leonard-ct**
As requested by https://github.com/systemd/systemd/issues/32027, this PR allows vmspawn to generate predictable TAP device names and MAC addresses with relatively little effort.
QEMU is still responsible for creating and destroying the devices and handling all the networking in between but now the network interface names are predicatable, following the same behaviour as nspawn, though without the fallback to truncation.

I had a bit of a think about a way to --network-tap-extra= but I can't see a simple way to do it with just QEMU arguments right now.

### CODE REVIEW

File: src/vmspawn/vmspawn.c [hunk 1]
```diff
@@ -1173,6 +1198,74 @@ static int generate_ssh_keypair(const char *key_path, const char *key_type) {
         return 0;
 }
 
+static int generate_mac(const char *machine_name, struct ether_addr *mac, sd_id128_t hash_key, uint64_t idx) {
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: I guess this is basically copied from nspawn-network.c. Maybe, better to expose it to e.g. netif-util.[ch] and reuse it.

File: src/vmspawn/vmspawn.c [hunk 2]
```diff
@@ -1173,6 +1198,74 @@ static int generate_ssh_keypair(const char *key_path, const char *key_type) {
...
+
+        assert_cc(ETH_ALEN <= sizeof(result));
+        memcpy(mac->ether_addr_octet, &result, ETH_ALEN);
+
+        ether_addr_mark_random(mac);
+
+        return 0;
+}
+
+static int shorten_ifname(char *ifname) {
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Also this one.

File: src/vmspawn/vmspawn.c [hunk 3]
```diff
@@ -534,6 +554,10 @@ static int parse_argv(int argc, char *argv[]) {
                 arg_settings_mask |= SETTING_START_MODE;
         }
 
+        r = parse_environment();
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Currently, only the mac address is parsed and no relevant argument for this. So, that's not a big matter, but we should parse environment variable first, then parse arguments. So, please call this before `parse_argv()`.

File: src/vmspawn/vmspawn.c [hunk 4]
```diff
@@ -186,6 +192,20 @@ static int help(void) {
         return 0;
 }
 
+static int parse_environment(void) {
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c:198]: Or, rename this to more specific, e.g. parse_mac_address_from_env(), and call before generating MAC address.

**Review by yuwata — CHANGES_REQUESTED** [review]
Please split the commit into small parts, 
- several cleanups (sorting headers, inserting missing empty line, manpage updates)
- move generate_mac() and friends from nspawn to neif-util.c,
- then, introduce persistent name and mac address to vmspawn.

**Comment by klausenbusk** [comment]
I just tested this on top of bffc1a28d50b3491e473e375b239e82bb7c5f419 and noticed that the MAC address is set to `00:00:00:00:00:00`.

Tested with `sudo systemd-vmspawn --image=image.raw --network-tap -M foo` and `-nic tap,ifname=tp-foo,script=no,model=virtio-net-pci,mac=00:00:00:00:00:00` is passed to QEMU.

**Comment by sam-leonard-ct** [comment]
> I just tested this on top of [bffc1a2](https://github.com/systemd/systemd/commit/bffc1a28d50b3491e473e375b239e82bb7c5f419) and noticed that the MAC address is set to `00:00:00:00:00:00`.
> 
> Tested with `sudo systemd-vmspawn --image=image.raw --network-tap -M foo` and `-nic tap,ifname=tp-foo,script=no,model=virtio-net-pci,mac=00:00:00:00:00:00` is passed to QEMU.

Thanks for flagging this up I will look into it.

**Comment by sam-leonard-ct** [comment]
@klausenbusk I believe I have now fixed this bug.

I was writing the generated mac into the wrong variable...

File: src/nspawn/nspawn-network.c [hunk 5]
```diff
@@ -8,6 +8,7 @@
 #include <sys/file.h>
 #include <sys/mount.h>
 
+#include "netif-util.h"
```
Comment (yuwata) [inline src/nspawn/nspawn-network.c]: Oh, you put the header in a wrong place. Please move this below.

File: src/vmspawn/vmspawn.c [hunk 6]
```diff
@@ -8,6 +10,7 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
+#include "netif-util.h"
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Please move this to the list of our internal headers.

File: src/vmspawn/vmspawn.c [hunk 7]
```diff
@@ -1287,14 +1308,32 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
-                r = strv_extend_many(&cmdline, "-nic", "user,model=virtio-net-pci");
-        else
-                r = strv_extend_many(&cmdline, "-nic", "none");
-        if (r < 0)
-                return log_oom();
+        if (arg_network_stack == NETWORK_STACK_TAP) {
+                _cleanup_free_ char *tap_name = NULL;
+                struct ether_addr mac_vm = {};
+
+                tap_name = strjoin("tp-", arg_machine);
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c:1315]: Missing OOM check.

File: man/systemd-vmspawn.xml [hunk 8]
```diff
@@ -498,6 +498,14 @@
     <title>Environment</title>
 
     <variablelist class='environment-variables'>
+      <varlistentry>
+        <term><varname>$SYSTEMD_VMSPAWN_NETWORK_MAC</varname></term>
```
Comment (yuwata) [inline man/systemd-vmspawn.xml]: Hmm, is this necessary? If necessary, why not to make this a regular argument? E.g. `--network-tap-mac=` or so.

File: src/vmspawn/vmspawn.c [hunk 9]
```diff
@@ -1287,14 +1308,32 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
+                        return log_oom();
+
+                r = strv_extendf(&cmdline, "tap,ifname=%s,script=no,model=virtio-net-pci,mac=%s", tap_name, ETHER_ADDR_TO_STR(&mac_vm));
+                if (r < 0)
+                        return log_oom();
+        } else {
+                r = strv_extend_many(&cmdline, "-nic", arg_network_stack == NETWORK_STACK_USER ? "user,model=virtio-net-pci" : "none");
+                if (r < 0)
+                        return log_oom();
+        }
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Not a matter, but I prefer previous style:
```c
                r = strv_extendf(&cmdline, "tap,ifname=%s,script=no,model=virtio-net-pci,mac=%s", tap_name, ETHER_ADDR_TO_STR(&mac_vm));
        } else if (arg_network_stack == NETWORK_STACK_USER)
                r = strv_extend_many(&cmdline, "-nic", "user,model=virtio-net-pci");
        else
                r = strv_extend_many(&cmdline, "-nic", "none");
        if (r < 0)
                return log_oom();
```

File: src/vmspawn/vmspawn.c [hunk 10]
```diff
@@ -1287,14 +1308,32 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
-        else
-                r = strv_extend_many(&cmdline, "-nic", "none");
-        if (r < 0)
-                return log_oom();
+        if (arg_network_stack == NETWORK_STACK_TAP) {
+                _cleanup_free_ char *tap_name = NULL;
+                struct ether_addr mac_vm = {};
+
+                tap_name = strjoin("tp-", arg_machine);
+                (void) net_shorten_ifname(tap_name);
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Please make it take additional flag to conditionalize `naming_scheme_has(NAMING_NSPAWN_LONG_HASH)`, as it is for nspawn. For vmspawn, we should always hash long name, I think.

File: src/vmspawn/vmspawn.c [hunk 11]
```diff
@@ -65,6 +69,8 @@
 #include "vmspawn-settings.h"
 #include "vmspawn-util.h"
 
+#define VM_HASH_KEY SD_ID128_MAKE(01,d0,c6,4c,2b,df,24,fb,c0,f8,b2,09,7d,59,b2,93)
```
Comment (yuwata) [inline src/vmspawn/vmspawn.c]: Please use more specific name, e.g. VM_TAP_HASH_KEY or so.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/systemd-vmspawn.xml (hunk 8 above)
Comment (sam-leonard-ct) [reply man/systemd-vmspawn.xml]: I did it like this to mirror what nspawn supports, I'd be happy to add an argument for it :)

Comment (yuwata) [reply man/systemd-vmspawn.xml]: Oh, I did not know nspawn supports similar env var.
Then, please document it in docs/ENVIRONMENT.md, and hide it from the man page, at least until we officially expose the one for nspawn.

Comment (sam-leonard-ct) [reply man/systemd-vmspawn.xml]: I didn't either until I read the code, I'll move it to the right place and revert my previous man page change splitting out environment variables.

**Review by yuwata — APPROVED** [review]

**Pull request merged**